### PR TITLE
Export a helper for rendering progress to a "simple" terminal

### PR DESCRIPTION
**Pull request by AaronFriel**
The helper method exported here uses an internal terminal, "SimpleTerminal", which ignores terminal control sequences and writes its output directly to a buffer. The wrapper retains a similar API as `ShowProgressEvents` to enable streaming events, though performance may be suboptimal as a subset of events must be buffered as rows to render the tree.

For cleaner output in this mode, the `CarriageReturn` operation is added to the internal terminal interface to conditionally render it. Ordinarily this is used along with cursor movement to render the tree in place, and is not needed for this mode. This removes spurious `\r` (rendered as "^M") characters from the diff for the plain rendering tests.

### CODE REVIEW

File: pkg/backend/display/testdata/not-truncated/failed-stacks-dont-rewrite-operations.json.interactive-100x80-plain.txt [hunk 1]
```diff
@@ -1,10 +1,10 @@
 View in Browser (Ctrl+O): link
 
-<%hide-cursor%>     Type                               Name      Status                  Info<%clear-to-end%>
-     pulumi:pulumi:Stack                test-dev                          1 mes<%clear-to-end%>
- --  ├─ pulumi-python:dynamic:Resource  test-r    deleted original        <%clear-to-end%>
- ~   └─ pulumi-python:dynamic:Resource  test-x    **updating failed**     <%clear-to-end%>
-<%clear-to-end%>
+     Type                               Name      Status                  Info
```
Comment (AaronFriel) [inline pkg/backend/display/testdata/not-truncated/failed-stacks-dont-rewrite-operations.json.interactive-100x80-plain.txt:3]: Not shown: there is an `\r` in the raw diff that was removed. This character is visible via `git diff` or in an editor that shows control characters.

File: pkg/backend/display/progress.go [hunk 2]
```diff
@@ -274,6 +274,61 @@ func ShowProgressEvents(op string, action apitype.UpdateKind, stack tokens.Stack
...
+//
+// A "simple" terminal is used which does not render control sequences. The simple terminal's output
+// is written to opts.Stdout.
+//
+// For consistent output approximating a terminal, consider setting:
+//
+//	opts.Color = colors.Never
+//	opts.RenderOnDirty = false
+//	opts.ShowResourceChanges = true
+//	opts.IsInteractive = true
```
Comment (pgavlin) [inline pkg/backend/display/progress.go:288]: should we force some or all of these options? without them, we'll still get a series of frames, yes?

Comment (AaronFriel) [reply pkg/backend/display/progress.go:288]: We could, think we ought to? It does simplify the contract with the caller if we set color, render on dirty, and is interactive.

`ShowResourceChanges` was added to the list inadvertently, it isn't a prerequisite.

Comment (AaronFriel) [reply pkg/backend/display/progress.go]: Updated

File: pkg/backend/display/progress.go [hunk 3]
```diff
@@ -274,6 +274,63 @@ func ShowProgressEvents(op string, action apitype.UpdateKind, stack tokens.Stack
...
+	action apitype.UpdateKind,
+	stack tokens.StackName,
+	proj tokens.PackageName,
+	permalink string,
+	events <-chan engine.Event,
+	done chan<- bool,
+	opts Options,
+	isPreview bool,
+	width, height int,
+) {
```
Comment (pgavlin) [inline pkg/backend/display/progress.go:299]: I'm kind of wondering if we should go even further.

Should we remove the `opts` parameter entirely and just have this method return a string?

Nice to be able to take a `writer`--that leaves the possibility open for streaming--but do we even need any of the other options?

I'm slightly uncomfortable with the name, but I don't have a better suggestion. I think that the docs are likely sufficient.

File: pkg/backend/display/progress.go [hunk 4]
```diff
@@ -274,6 +274,63 @@ func ShowProgressEvents(op string, action apitype.UpdateKind, stack tokens.Stack
...
+	o := opts
+	if o.term == nil {
+		o.term = terminal.NewSimpleTerminal(o.Stdout, width, height)
+	}
```
Comment (pgavlin) [inline pkg/backend/display/progress.go:303]: should we just override the terminal entirely s.t. it's always a `SimpleTerminal`?

**Review by pgavlin — APPROVED** [review]
Two more questions, but I'm okay with this as-is as well.

File: pkg/backend/display/progress_test.go [hunk 5]
```diff
@@ -154,13 +154,8 @@ func TestProgressEvents(t *testing.T) {
 					t.Run("plain", func(t *testing.T) {
 						suffix := fmt.Sprintf(".interactive-%vx%v-plain", width, height)
 						opts := defaultOpts()
-						opts.Color = colors.Never
-						opts.RenderOnDirty = false
 						opts.ShowResourceChanges = true
-						opts.IsInteractive = true
-						opts.ShowResourceChanges = true
-						testProgressEvents(t, path, accept, suffix, opts, width, height, true)
```
Comment (Frassle) [inline pkg/backend/display/progress_test.go:162]: Why is it ok to replace these tests with just the simpleRenderer?

File: pkg/backend/display/progress.go (hunk 3 above)
Comment (Frassle) [reply pkg/backend/display/progress.go:299]: Inclined to agree with removing the `opts` if we're not respecting most of them.

File: pkg/backend/display/progress_test.go (hunk 5 above)
Comment (AaronFriel) [reply pkg/backend/display/progress_test.go:162]: These tests were added by me for this purpose in another recent PR.

Comment (Frassle) [reply pkg/backend/display/progress_test.go:162]: Ah ok 👍

**Review by Frassle — APPROVED** [review]
Looks ok to me. Agree with Pat about maybe removing `opts` but could go either way.

**Pull request merged**